core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java

```diff
@@ -3647,6 +3720,211 @@ public Node visitJsonTableDefaultPlan(SqlBaseParser.JsonTableDefaultPlanContext
+    }
+
+    @Override
+    public Node visitLanguageCharacteristic(SqlBaseParser.LanguageCharacteristicContext context)
+    {
+        return new LanguageCharacteristic(getLocation(context), (Identifier) visit(context.identifier()));
+    }
+
+    @Override
+    public Node visitDeterministicCharacteristic(SqlBaseParser.DeterministicCharacteristicContext context)
+    {
+        return new DeterministicCharacteristic(getLocation(context), context.NOT() == null);
+    }
+
+    @Override
+    public Node visitReturnsNullOnNullInputCharacteristic(SqlBaseParser.ReturnsNullOnNullInputCharacteristicContext context)
+    {
+        return NullInputCharacteristic.returnsNullOnNullInput(getLocation(context));
+    }
+
+    @Override
+    public Node visitCalledOnNullInputCharacteristic(SqlBaseParser.CalledOnNullInputCharacteristicContext context)
+    {
+        return NullInputCharacteristic.calledOnNullInput(getLocation(context));
+    }
+
+    @Override
+    public Node visitSecurityCharacteristic(SqlBaseParser.SecurityCharacteristicContext context)
+    {
+        return new SecurityCharacteristic(getLocation(context), (context.INVOKER() != null)
+                ? SecurityCharacteristic.Security.INVOKER
+                : SecurityCharacteristic.Security.DEFINER);
+    }
+
+    @Override
+    public Node visitCommentCharacteristic(SqlBaseParser.CommentCharacteristicContext context)
+    {
+        return new CommentCharacteristic(getLocation(context), ((StringLiteral) visit(context.string())).getValue());
+    }
+
+    @Override
+    public Node visitReturnStatement(SqlBaseParser.ReturnStatementContext context)
+    {
+        return new ReturnStatement(getLocation(context), (Expression) visit(context.valueExpression()));
+    }
+
+    @Override
+    public Node visitAssignmentStatement(SqlBaseParser.AssignmentStatementContext context)
+    {
+        return new AssignmentStatement(
+                getLocation(context),
+                (Identifier) visit(context.identifier()),
+                (Expression) visit(context.expression()));
+    }
+
+    @Override
+    public Node visitSimpleCaseStatement(SqlBaseParser.SimpleCaseStatementContext context)
+    {
+        return new CaseStatement(
+                getLocation(context),
+                visitIfPresent(context.expression(), Expression.class),
+                visit(context.caseStatementWhenClause(), CaseStatementWhenClause.class),
+                visitIfPresent(context.elseClause(), ElseClause.class));
+    }
+
+    @Override
+    public Node visitSearchedCaseStatement(SqlBaseParser.SearchedCaseStatementContext context)
+    {
+        return new CaseStatement(
+                getLocation(context),
+                Optional.empty(),
+                visit(context.caseStatementWhenClause(), CaseStatementWhenClause.class),
+                visitIfPresent(context.elseClause(), ElseClause.class));
+    }
```

Comment:
These need to be separate constructs. The simple case statement needs to be able to handle stuff like:

```
CASE x
    WHEN > 5 THEN ...
    WHEN IS NULL THEN ...
    ...
END CASE
```

Reply:
Interesting, though we don't support that today for CASE expressions.